source/dhtslib/threadpool.d

```diff
@@ -0,0 +1,482 @@
+// but this time use a function pointer instead
+// of an alias to represent std.file.read.
+import std.file;
+void main()
+{
+    // Create and execute a Task for reading
+    // foo.txt.
+    auto file1Task = task(&read!string, "foo.txt", size_t.max);
+    file1Task.executeInNewThread();
+    // Read bar.txt in parallel.
+    auto file2Data = read("bar.txt");
+    // Get the results of reading foo.txt.
+    auto file1Data = file1Task.yieldForce;
+}
+---
+Notes: This function takes a non-scope delegate, meaning it can be
+       used with closures.  If you can't allocate a closure due to objects
+       on the stack that have scoped destruction, see `scopedTask`, which
+       takes a scope delegate.
+ */
+auto task(F, Args...)(F delegateOrFp, Args args)
+if (is(typeof(delegateOrFp(args))) && !isSafeTask!F)
+{
+    alias T = Task!(run, F, Args);
+    auto task = cast(T*) calloc(T.sizeof, 1);
+    *task = T(delegateOrFp, args);
+    return task;
+}
+
+/**
+Version of `task` usable from `@safe` code.  Usage mechanics are
+identical to the non-@safe case, but safety introduces some restrictions:
+1.  `fun` must be @safe or @trusted.
+2.  `F` must not have any unshared aliasing as defined by
+    $(REF hasUnsharedAliasing, std,traits).  This means it
+    may not be an unshared delegate or a non-shared class or struct
+    with overloaded `opCall`.  This also precludes accepting template
+    alias parameters.
+3.  `Args` must not have unshared aliasing.
+4.  `fun` must not return by reference.
+5.  The return type must not have unshared aliasing unless `fun` is
+    `pure` or the `Task` is executed via `executeInNewThread` instead
+    of using a `TaskPool`.
+*/
+@trusted auto task(F, Args...)(F fun, Args args)
+if (is(typeof(fun(args))) && isSafeTask!F)
+{
+    alias T = Task!(run, F, Args);
+    auto task = cast(T*) calloc(T.sizeof, 1);
+    *task = T(fun, args);
+    return task;
+}
+
+debug(dhtslib_unittest) unittest {
+    auto add = (int a, int b) => a + b;
+    auto t = add.task(3, 4);
+    t = cast(typeof(t))t.runTask(t);
+    assert(t.returnVal == 7);
+    import std.exception : assertThrown;
+    assertThrown(t.enforcePool);
+}
+
+/// End of Boost Licensed code
+
+/// Wraps an htslib hts_tpool_process into a range interface. Acts as a parallel forward range.
+struct Process(alias fun, R)
+{
+    ThreadPool * pool;
+
+    HtsProcess proc;
+
+    alias Args = ElementType!R;
+
+    R range;
+
+    alias TaskType = PointerTarget!(ReturnType!(task!(fun, Args)));
+    
+
+    TaskType frontVal;
+
+    bool end;
+    bool empty;
+    
+    this(R range, ThreadPool * pool) {
+        this.pool = pool;
+        this.range = range;
+        this.proc = HtsProcess(hts_tpool_process_init(this.pool.tpool, cast(int)this.pool.threads * 2, 0));
+        frontVal = frontVal.init;
+        auto qsize =  hts_tpool_process_qsize(this.proc);
+        for(auto i = 0; i < qsize; i++) {
+            if(this.range.empty) {
+                break;
```

Comment:
empty range condition seems important to test -- why do yoyu think it was not covered?

Reply:
I simply did not add a test for an empty range. Will add.